tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,167 @@
+		logger: logger,
+		dryRun: dryRun,
+	}, err
+}
+
+// Run iterates over all blocks from all users in a bucket
+func (c ThanosBlockConverter) Run(ctx context.Context) (Results, error) {
+	results := make(Results)
+
+	scanner := cortex_tsdb.NewUsersScanner(c.bkt, cortex_tsdb.AllUsers, c.logger)
+	users, _, err := scanner.ScanUsers(ctx)
+	if err != nil {
+		return results, errors.Wrap(err, "error while scanning users")
+	}
+	level.Info(c.logger).Log("msg", "Scanned users")
+
+	for _, u := range users {
+		r, err := c.convertUser(ctx, u)
+		results[u] = r
+		if err != nil {
+			return results, errors.Wrap(err, fmt.Sprintf("error converting user %s", u))
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	userBucketClient := bucket.NewUserBucketClient(user, c.bkt)
+	err := userBucketClient.Iter(ctx, "", func(o string) error {
+		blockID, ok := block.IsBlockDir(o)
+		if !ok {
+			// not a block
+			return nil
+		}
+
+		// retrieve meta.json
+
+		meta, err := block.DownloadMeta(ctx, c.logger, userBucketClient, blockID)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "download block meta", "block", blockID.String(), "err", err.Error())
+			results.AddFailed(blockID.String())
+			return nil
+		}
+
+		// convert and upload if appropriate
+
+		newMeta, changesRequired := convertMetadata(meta, user)
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Info(c.logger).Log("msg", "Block requires changes (dry-run)", "block", blockID.String(), "changes_required", strings.Join(changesRequired, ","))
+			} else {
+				level.Info(c.logger).Log("msg", "Block requires changes, uploading new meta.json", "block", blockID.String(), "changes_required", strings.Join(changesRequired, ","))
+				if err := c.uploadNewMeta(ctx, userBucketClient, blockID.String(), newMeta); err != nil {
+					level.Error(c.logger).Log("msg", "Update meta.json", "block", blockID.String(), "err", err.Error())
+					results.AddFailed(blockID.String())
+					return nil
+				}
+			}
+			results.AddConverted(blockID.String())
+		} else {
+			level.Info(c.logger).Log("msg", "Block doesn't need changes", "block", blockID.String())
+			results.AddUnchanged(blockID.String())
+		}
+
+		return nil
+	})
+	if err != nil {
+		return results, err
+	}
+
+	return results, nil
+}
+
+func (c *ThanosBlockConverter) uploadNewMeta(ctx context.Context, userBucketClient *bucket.UserBucketClient, blockID string, meta metadata.Meta) error {
+	var body bytes.Buffer
+	if err := meta.Write(&body); err != nil {
+		return errors.Wrap(err, "encode json")
+	}
+
+	if err := userBucketClient.Upload(ctx, fmt.Sprintf("%s/meta.json", blockID), &body); err != nil {
+		return errors.Wrap(err, "upload to bucket")
+	}
+	return nil
+}
+
+func convertMetadata(meta metadata.Meta, expectedUser string) (metadata.Meta, []string) {
+	var changesRequired []string
+
+	if meta.Thanos.Labels == nil {
```

Comment:
No need for this check or creating of empty `meta.Thanos.Labels`, since it will be replaced anyway. Reading from `nil` map, or deleting from `nil` map will work as expected. (I'd suggest adding unit test for that)